lib/stats/stats-cluster-key-builder.h

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright (c) 2023 Attila Szakacs
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#ifndef STATS_CLUSTER_KEY_BUILDER_H_INCLUDED
+#define STATS_CLUSTER_KEY_BUILDER_H_INCLUDED
+
+#include "syslog-ng.h"
+#include "stats-cluster.h"
+
+typedef struct _StatsClusterKeyBuilder StatsClusterKeyBuilder;
+
+StatsClusterKeyBuilder *stats_cluster_key_builder_new(void);
+void stats_cluster_key_builder_free(StatsClusterKeyBuilder *self);
+
+void stats_cluster_key_builder_set_name(StatsClusterKeyBuilder *self, const gchar *name);
+void stats_cluster_key_builder_set_name_suffix(StatsClusterKeyBuilder *self, const gchar *name_suffix);
+void stats_cluster_key_builder_add_label(StatsClusterKeyBuilder *self, const StatsClusterLabel label);
+void stats_cluster_key_builder_set_unit(StatsClusterKeyBuilder *self, StatsClusterUnit unit);
+void stats_cluster_key_builder_set_legacy_alias(StatsClusterKeyBuilder *self, guint16 component, const gchar *id,
+                                                const gchar *instance);
+void stats_cluster_key_builder_set_legacy_alias_name(StatsClusterKeyBuilder *self, const gchar *name);
+
+void stats_cluster_key_builder_build_single(StatsClusterKeyBuilder *self, StatsClusterKey *sc_key);
```

Comment:
Suggestion:
Why don't we separate cheap and more expensive key creations?

Let's say that stack-based key allocation is cheap, and the KeyBuilder should not be used in such cases (the builder is heap-allocated, it can be considered more expensive: you wouldn't use it for dynamic counters).

Since we don't really have heap-allocated keys yet, we could say that the more expensive method should be used through the KeyBuilder.

Based on the above, we could change the signature of a few functions:
```
StatsClusterKey *stats_cluster_key_builder_build_single(StatsClusterKeyBuilder *self);
StatsClusterKey *stats_cluster_key_builder_build_logpipe(StatsClusterKeyBuilder *self);
```
And then we could:
```diff
+ stats_cluster_key_new() // package-private method (only the builder should use it)
-stats_cluster_key_cloned_free()
+stats_cluster_key_free() // package-private method (only the builder should use it)
```

Reply:
This is a really good idea.

I think `free()` must be used by the caller, to free the created key.

And if I understand correctly `new()` would be a `g_new0()` call only, which can be done in the `build()` functions, too, so we do not expose something, which is not meant to be used.

If these are all true, then this task reduces to a function rename from `stats_cluster_key_cloned_free()` to `stats_cluster_key_free()`, and allocating the key inside `build()` and returning that.

Reply:
Oh `stats_cluster_key_free()` needs to free `self`, too, so we need both, I think.

Reply:
Implemented this, but we can iterate on it further.

Reply:
LGTM.